docs/blog/12factor.md

```diff
@@ -0,0 +1,219 @@
+
+Fulfilling this factor on an existing code base is much harder than it sounds. It means mapping all the (often implicit) state machines involved in the app and coordinating them so that there are no undefined transitions. For example, the database connection pool must be ready before we bring up our HTTP listener and must not shut down until the HTTP listener is down _and_ all in-flight HTTP requests are done. Workers must "hand back" in-progress work items so that replacement workers do not have to wait for timeout to process those work items. This difficulty is compounded with distributed systems as a conceptual "transaction" may span more than one app or backing service (e.g. writing to file storage and sending mail), requiring [two-phase commit](https://en.wikipedia.org/wiki/Two-phase_commit_protocol) semantics.
+
+This factor is nevertheless, the key to the always-on experience that we take for granted in large cloud services.
+
+Indicators:
+
+- long delay or elevated error rate when deploying because load balancers misjudge the app's readiness
+- manual steps in deployment even when deployments are nominally automated
+- deploys only during off-hours
+
+## Factor X: Dev/prod parity
+
+> The twelve-factor app is designed for continuous deployment by keeping the gap between development and production small.
+
+This factor has still not established itself fully in that many production environments are hard to squeeze onto a developer's laptop. Generally speaking, the public cloud providers put too little effort into supporting development use cases for their services.
+
+[Kind](https://kind.sigs.k8s.io/) deserves mentioning for its heroic effort to achieve a dev-friendly, multi-node Kuberentes cluster using only Docker.
+
+Docker has introduced a borderland where it is possible to develop a container using just Docker Engine for dev environment and still be reasonably confident that it will execute properly in e.g. Kubernetes. Still, some provisioning of backend services is still needed and time is wasted maintaining two different sets of instrumentation. For example, apps often have a Docker Compose file to get developers started, and a Kubernetes manifest for test/prod.
+
+> The twelve-factor developer resists the urge to use different backing services between development and production
+
+Many full-stack development setup includes "dev" servers whose role is to automatically reload the app as its source code change. Similarly, tools like Docker Desktop and Tilt provide capabilities and environments that are subtly different from e.g. Azure Container instances or Kubernetes. All these will color developers' choices and risk introducing issues that will not be discovered until late in the process.
```

Comment:
A link to Tilt and maybe a mention of Azure DevSpaces?

Reply:
Tilt, indeed. Devspaces https://azure.microsoft.com/en-us/updates/azure-dev-spaces-is-retiring-on-31-october-2023/ ?